### PR TITLE
[Enhancement] Add BaseSOTDataset, refactor GOT10kDataset and DistributedVideoSampler in order to support online annotation parsing in SOT dataset.

### PR DESCRIPTION
**Pull request by JingweiZhang12**
Some SOT methods like STARK use TrackingNet as training data. 

While MMTracking preprocesses the annotations of TrackingNet to a JSON file that is too large to be loaded during training.

Therefore, we turn to use online annotation parsing in the new sot dataset class `BaseSOTDataset` from now.

This PR mainly refactor 3 features:
1. Provide a new `BaseSOTDataset` in order to online parse annotations
2. Refactor `GOT10kDataset`, which inherits `BaseSOTDataset` class rather than `SOTTestDataset` class, as an example of how to develop a SOT dataset by inheriting `BaseSOTDataset`.
3. We need pass an `tuple(video_id, frame_id)` to the `__getitem__()` function, in testing mode of `BaseSOTDataset` and `GOT10kDataset`. Therefore we also refactor the `DistributedVideoSampler` to support SOT dataset testing.

### CODE REVIEW

File: mmtrack/datasets/__init__.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 from .parsers import CocoVID
 from .pipelines import PIPELINES
 from .reid_dataset import ReIDDataset
+from .sot_dataset import SOTDataset
```
Comment (GT9505) [inline mmtrack/datasets/__init__.py:14]: `from .base_sot_dataset import BaseSOTDataset`

File: mmtrack/datasets/sot_dataset.py [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    """
+
+    CLASSES = None
+
+    def __init__(self, img_prefix, pipeline, split, test_mode=False, **kwargs):
+        self.img_prefix = img_prefix
+        self.split = split
+        self.pipeline = Compose(pipeline)
+        self.test_mode = test_mode
+        self.is_video_dataset = True
```
Comment (GT9505) [inline mmtrack/datasets/sot_dataset.py:28]: Why do you always  need to set `is_video_dataset=True`

File: mmtrack/datasets/sot_dataset.py [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        self.bbox_min_size = 0
```
Comment (GT9505) [inline mmtrack/datasets/sot_dataset.py:29]: change `bbox_min_size` to an argument

File: mmtrack/datasets/sot_dataset.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+            number of videos. The default content is in the following format:
+            [
+                {
+                    'video_path': the video path
+                    'ann_path': the annotation path
+                    'start_frame': the starting frame number contained in the
+                                    image name
+                    'end_frame': the ending frame number contained in the image
+                                    name
+                    'num_digit': the number of digits in the image name
```
Comment (GT9505) [inline mmtrack/datasets/sot_dataset.py:40]: change `num_digit` to `framename_template`

File: mmtrack/datasets/sot_dataset.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+        self.test_mode = test_mode
+        self.is_video_dataset = True
+        self.bbox_min_size = 0
+        ''' The self.data_info is a list, which the length is the
+            number of videos. The default content is in the following format:
+            [
+                {
+                    'video_path': the video path
+                    'ann_path': the annotation path
+                    'start_frame': the starting frame number contained in the
```
Comment (GT9505) [inline mmtrack/datasets/sot_dataset.py:36]: `start_frame_id`

File: mmtrack/datasets/sot_dataset.py [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+        '''
+        self.data_infos = self.load_data_infos(split=self.split)
+        self.num_frames_per_video = []
+        for video_ind in range(len(self.data_infos)):
+            self.num_frames_per_video.append(self.get_len_per_video(video_ind))
+
+    def __getitem__(self, ind):
+        if self.test_mode:
+            assert isinstance(ind, tuple)
+            return self.prepare_test_img(ind[0], ind[1])
```
Comment (GT9505) [inline mmtrack/datasets/sot_dataset.py:53]: add comment to explain what the `tuple` represents

File: mmtrack/datasets/sot_dataset.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+                                    name
+                    'num_digit': the number of digits in the image name
+                },
+                ...
+            ]
+        '''
+        self.data_infos = self.load_data_infos(split=self.split)
+        self.num_frames_per_video = []
+        for video_ind in range(len(self.data_infos)):
+            self.num_frames_per_video.append(self.get_len_per_video(video_ind))
```
Comment (GT9505) [inline mmtrack/datasets/sot_dataset.py:48]: `self.num_frames_per_video = [self.get_len_per_video(video_ind) for video_ind in range(len(self.data_infos))]`

File: mmtrack/datasets/sot_dataset.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+from mmtrack.core.evaluation import eval_sot_ope
+from mmtrack.datasets import DATASETS
+
+
+@DATASETS.register_module()
+class SOTDataset(Dataset, metaclass=ABCMeta):
+    """Dataset of single object tracking.
+
+    The dataset can both support training and testing mode.
+    """
```
Comment (GT9505) [inline mmtrack/datasets/sot_dataset.py:19]: add docstring for all functions

File: mmtrack/datasets/sot_dataset.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+            return self.prepare_train_img(ind)
+
+    @abstractmethod
+    def load_data_infos(self, split='train'):
+        pass
+
+    def get_img_names_from_video(self, video_ind):
+        img_names = []
+        start_frame, end_frame = self.data_infos[video_ind][
+            'start_frame'], self.data_infos[video_ind]['end_frame']
```
Comment (GT9505) [inline mmtrack/datasets/sot_dataset.py:64]: split this line to 2 lines

File: mmtrack/datasets/sot_dataset.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+        if len(bboxes.shape) == 1:
+            bboxes = np.expand_dims(bboxes, axis=0)
+
+        end_frame, start_frame = self.data_infos[video_ind][
+            'end_frame'], self.data_infos[video_ind]['start_frame']
+
+        if not self.test_mode:
+            assert len(bboxes) == (
+                end_frame - start_frame +
+                1), f'{len(bboxes)}, {end_frame}, {start_frame}'
```
Comment (GT9505) [inline mmtrack/datasets/sot_dataset.py:87]: complete the error information

File: mmtrack/datasets/sot_dataset.py [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+        return bboxes
```
Comment (GT9505) [inline mmtrack/datasets/sot_dataset.py:88]: describe the format of `bboxes` in the docstring

File: mmtrack/datasets/sot_dataset.py [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+
+    def get_masks_from_video(self, video_ind):
+        pass
+
+    def get_ann_infos_from_video(self, video_ind):
+        bboxes = self.get_bboxes_from_video(video_ind)
+        visible_info = self.get_visibility_from_video(video_ind)
+        bboxes_isvalid = (bboxes[:, 2] > self.bbox_min_size) & (
+            bboxes[:, 3] > self.bbox_min_size)
+        visible_info['visible'] = visible_info['visible'] & bboxes_isvalid
```
Comment (GT9505) [inline mmtrack/datasets/sot_dataset.py:106]: I remember that we should not merge `visible` and `bbox_isvalid` in here

Comment (JingweiZhang12) [reply mmtrack/datasets/sot_dataset.py:106]: It follows the `pytracking`. Generally, we think the object with 'visible'=True can be sampled directly, so it's annotation must valid and we merge `visible` and `bbox_isvalid`.

File: mmtrack/datasets/base_sot_dataset.py [hunk 13]
```diff
@@ -0,0 +1,285 @@
...
+from mmtrack.core.evaluation import eval_sot_ope
+from mmtrack.datasets import DATASETS
+
+
+@DATASETS.register_module()
+class SOTDataset(Dataset, metaclass=ABCMeta):
+    """Dataset of single object tracking.
+
+    The dataset can both support training and testing mode.
+    """
```
Comment (GT9505) [inline mmtrack/datasets/base_sot_dataset.py:19]: complete the docstring for the `class`. you can refer to this [example](https://github.com/open-mmlab/mmtracking/blob/master/mmtrack/models/trackers/masktrack_rcnn_tracker.py#L14)

File: mmtrack/datasets/base_sot_dataset.py [hunk 14]
```diff
@@ -0,0 +1,285 @@
...
+                 bbox_min_size=0,
+                 load_as_video=True,
+                 **kwargs):
+        self.img_prefix = img_prefix
+        self.split = split
+        self.pipeline = Compose(pipeline)
+        self.test_mode = test_mode
+        self.bbox_min_size = bbox_min_size
+        # for DistributedVideoSampler in testing
+        self.laod_as_video = load_as_video
```
Comment (GT9505) [inline mmtrack/datasets/base_sot_dataset.py:37]: fix the typo

File: mmtrack/datasets/base_sot_dataset.py [hunk 15]
```diff
@@ -0,0 +1,285 @@
...
+
+
+@DATASETS.register_module()
+class SOTDataset(Dataset, metaclass=ABCMeta):
+    """Dataset of single object tracking.
+
+    The dataset can both support training and testing mode.
+    """
+
+    CLASSES = None
```
Comment (GT9505) [inline mmtrack/datasets/base_sot_dataset.py:21]: Add comments to explain why we set `CLASSES = None`.
For any weird variables, we need to add comments to explain the motivation.

File: mmtrack/datasets/base_sot_dataset.py [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+        """
+        img_names = []
+        start_frame_id = self.data_infos[video_ind]['start_frame_id']
+        end_frame_id = self.data_infos[video_ind]['end_frame_id']
+        framename_template = self.data_infos[video_ind]['framename_template']
+        for frame_id in range(start_frame_id, end_frame_id + 1):
+            img_names.append(
+                osp.join(
+                    self.data_infos[video_ind]['video_path'],
+                    '{frame_id:0{framename_template}}.jpg'.format(
```
Comment (GT9505) [inline mmtrack/datasets/base_sot_dataset.py:90]: The `framename_template` shoud not be an int. We can set `framename_template = '%06d.jpg'` in a specifical dataset , and use `framename_template % frame_id` to generate the filename. You can give it a try.

File: mmtrack/datasets/base_sot_dataset.py [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+                ann_infos = self.get_ann_infos_from_video(video_index)
+                img_infos = self.get_img_infos_from_video(video_index)
+                video_infos = dict(**ann_infos, **img_infos)
+                self.pre_pipeline(video_infos)
+                pair_video_infos.append(video_infos)
+
+            results = self.pipeline(pair_video_infos)
+            if results is None:
+                continue
+            return results
```
Comment (GT9505) [inline mmtrack/datasets/base_sot_dataset.py:220]: the 3 lines above can be changed to
`if results is not None: return results`

File: mmtrack/datasets/got10k_dataset.py [hunk 18]
```diff
@@ -37,38 +141,31 @@ def format_results(self, results, resfile_path=None):
...
                     bbox = [
                         str(f'{bbox[0]:.4f}'),
                         str(f'{bbox[1]:.4f}'),
                         str(f'{(bbox[2] - bbox[0]):.4f}'),
                         str(f'{(bbox[3] - bbox[1]):.4f}')
                     ]
                     line = ','.join(bbox) + '\n'
-                    f.writelines(line)
+                    f_bbox.writelines(line)
+                    f_time.writelines('0.04\n')
```
Comment (GT9505) [inline mmtrack/datasets/got10k_dataset.py:167]: remove the time cost, since it has not been tested

File: mmtrack/datasets/got10k_dataset.py [hunk 19]
```diff
@@ -1,21 +1,125 @@
...
 from mmdet.datasets import DATASETS
 
-from .sot_test_dataset import SOTTestDataset
+from .base_sot_dataset import SOTDataset
 
 
 @DATASETS.register_module()
-class GOT10kDataset(SOTTestDataset):
-    """GOT10k dataset for the testing of single object tracking.
+class GOT10kDataset(SOTDataset):
```
Comment (GT9505) [inline mmtrack/datasets/got10k_dataset.py:15]: add `sampler` in order to test on GOT-10k dataset

**Comment by GT9505** [comment]
***Please note that we need to guarantee that the testing of `siameserpn++` can be running normally in the following pr which will change the `LaSOT`, `TrackingNet` dataset, ....*

File: mmtrack/datasets/base_sot_dataset.py (hunk 16 above)
Comment (JingweiZhang12) [reply mmtrack/datasets/base_sot_dataset.py:90]: It doesn't work. May be we can rename `framename_template` as 'num_digit_in_framename'?

File: mmtrack/datasets/got10k_dataset.py (hunk 18 above)
Comment (JingweiZhang12) [reply mmtrack/datasets/got10k_dataset.py:167]: The time cost must be written in the submitted files, otherwise, the testing server can not work.

Comment (JingweiZhang12) [reply mmtrack/datasets/got10k_dataset.py:167]: may be we can set `record_test_time=False` as a parameter in `format_results` function.

**Review by GT9505 — APPROVED** [review]

**Pull request merged**